### DISCUSSION TITLE
Input mode default to Hiragana, but keepig other Input sources preference order

### DISCUSSION THREAD

**Discussion by ryosuketc** [google/mozc#925] [Q&A]
Context: https://github.com/google/mozc/issues/381.

# Actual behavior:

When [active_on_launch](https://github.com/google/mozc/blob/master/docs/configurations.md#activate-mozc-on-launch) is enabled, it overrides the Settings > Keyboard > Input sources order.
For example, assuming that the two input sources are set: `English (US)` and `Japanese (Mozc)` and the `active_on_launch` is set to `True`, it changes the default input source to `Japanese (Mozc)` with Hiragana input mode selected (overriding the oder of `English (US)` -> `Japanese (Mozc)` in Settings). 

# Desired behavior:

I would like Mozc to change **only** the deafult **input mode for Mozc**, rather than overriding the input sources in Settings.
In the example above, I would like to keep `English (US)` as my first selection on system boot but Mozc to change the default input mode to Hiragata *when switched to `Japanese (Mozc)`*.

# Root cause:

The root cause of my ask here comes from the very first comment of https://github.com/google/mozc/issues/381 - basically Mozc selects `direct input` as the default input mode. I *assume* the most frequenty used mode is Hiragana, the default mode could be set as Hiragana, or at least it needs to be configurable if not yet. As I am somewhat new to Mozc, I may be just missing already-avalavile configuraiton options?

# Environments:
I am on Ubuntu 24.04 LTS.

**Comment by ryosuketc** [comment]
@yukawa thanks for the suggestion on moving to a new theread - @hiroyuki-komatsu for visibility.

**Comment by yukawa** [comment]
> The root cause of my ask here comes from the very first comment of https://github.com/google/mozc/issues/381 - basically Mozc selects `direct input` as the default input mode. I assume the most frequenty used mode is Hiragana, the default mode could be set as Hiragana, or at least it needs to be configurable if not yet. As I am somewhat new to Mozc, I may be just missing already-avalavile configuraiton options?

IIUC, Mozc `2.28.4950` and later do support such configuration.
https://github.com/google/mozc/blob/master/docs/configurations.md#fixed-composition-mode-per-engine
> ### Fixed composition mode per engine
> `composition_mode` specifies the composition mode every time the engine isenabled. This is useful to switch the composition mode by using Ibus hot keys.
>
> The available values are `DIRECT`, `HIRAGANA`, `FULL_KATAKANA`, `HALF_ASCII`, `FULL_ASCII`, `HALF_KATAKANA`, and `NONE`. `NONE` does not change the composition mode.
>
> `symbol` is a label to represent the engine used by Ibus.
>
> `composition_mode` and `symbol` are available in 2.28.4950 and later versions.
>```
>engines {
>  name : "mozc-on"
>  longname : "Mozc:あ"
>  layout : "default"
>  layout_variant : ""
>  layout_option : ""
>  rank : 80
>  symbol : "あ"
>  composition_mode: HIRAGANA
>}
>``` 

According to https://repology.org/project/mozc/versions, however, Ubuntu 24.04 shipped with `2.28.4715.102`, which does not have the above option yet unfortunately.

**Comment by hiroyuki-komatsu** [comment]
`active_on_launch` does not probably change the order of input sources.

I guess Ibus controls the order by using `rank`. Since the rank of `en_us` is apparently 50, would you check if it works when you set the rank of Mozc to a smaller number than 50 (e.g. 25)?

By the way, does it also work if you remove `en_us` from your input sources and use `Mozc` only?

**Comment by yukawa** [comment]
> In the example above, I would like to keep `English (US)` as my first selection on system boot but Mozc to change the default input mode to Hiragata when switched to `Japanese (Mozc)`.

Any chance that it's not Mozc but some other component, say GNOME, is trying to restore last used input source upon your logging in again? What would happen if you do the following?

 1. Have multiple IMEs in your input source, say IME X, IME Y, and IME Z in this order.
 2. Select IME Z
 3. Reboot your system
 4. See what IME is selected.

**Comment by ryosuketc** [comment]
It seems that changing the `rank` did the trick!

`~/.config/mozc/ibus_config.textproto`

```
engines {
  name : "mozc-jp"
  longname : "Mozc"
  layout : "default"
  layout_variant : ""
  layout_option : ""
  rank : 25 # Changed from the default 80.
}
active_on_launch: True # Added.
```
---
> By the way, does it also work if you remove `en_us` from your input sources and use `Mozc` only?

Yes. It seems that `active_on_launch` selects Hiragana on system boot when only Mozc exists in the input sources.

> Any chance that it's not Mozc but some other component, say Gnome, is trying to restore last used input source upon your logging in again? What would happen if you do the following?

I am not too sure how I could select multiple IMEs. 

Settings > Keyboard > Input sources: ![image](https://github.com/google/mozc/assets/43229670/2faff44f-d1ab-46bd-9e6d-85984ba92605)

Settings > Region & Language > Manage installed languages: ![image](https://github.com/google/mozc/assets/43229670/d19b374c-7316-4e38-9f1c-5d61dcd315b3)

From the second screenshot, I can select only one keyboard input method system, not one per an input source?

---

Anyways, it is working as I desire now, and thank you a lot @hiroyuki-komatsu and @yukawa for the investigation! I am closing this for now, but please let me know if you want me to test more to investigate / improve further.
